### ISSUE TITLE
Can't uninstall miniconda (after failed install): "Failed to run pre_uninstall"

### ISSUE THREAD

**Issue by morrisseyj** [ContinuumIO/anaconda-issues#13381]
### Checklist

- [X] I added a descriptive title.
- [X] I searched through [existing issues](https://github.com/ContinuumIO/anaconda-issues/issues) and couldn't find a solution or duplicate issue.
- [X] I searched on the web (e.g. Google) and didn't find any helpful information.
- [X] I searched the [Anaconda documentation](https://docs.anaconda.com/) and didn't find any helpful information.

### Impacted product

- [ ] A conda package (built by Anaconda)
- [ ] Anaconda Distribution (formerly Anaconda Individual Edition)
- [X] Miniconda
- [ ] Anaconda.org

### What happened?

On Windows 11. I am trying to uninstall miniconda. I run the uninstaller but get the message: "Failed to run pre_uninstall". Note this was after a failed install when I got a similar message during the installer. Searching online i saw that you should just ignore this message, during the install. I did that but the install failed - I could not access conda from the cmd line. Now i can' uninstall.

If, during the uninstall, I ignore the "Failed to run pre_uninstall" prompt, I get another error: "Failed to run rmpath". If i ignore that i get another one: "Failed to run rmreg". If i ignore that, the uninstaller completes. I finish and exit, however miniconda remains in my programs - even though its not working.

I can't run conda info/conda config as conda did not install.

How can i clean everything up to try and get things installed correctly?

### Expected behavior or outcome

Miniconda will uninstall.

**Comment by marcoesters** [comment]
Could you tell me which Miniconda installer you used and what the error during the installation was? Generally, those messages shouldn't be ignored because they point to actual problems.

There is a way to clean up the system manually. As with most Windows software, this is a little involved, but the steps are below. Depending on when the installation failed and which options you checked, some steps may not be necessary.

This assumes that you don't have any other Miniconda installation on the system before you tried to install it.

* Delete the directory you tried to install Miniconda into. 
* Type `%USERPROFILE%` into the Windows Exporer address field. You should see a `.conda` directory. If you do, open it and open the `environments.txt` file. Remove any lines that contain the directory of the failed Miniconda installation.
* Remove the start menu folder. To do that, open the Windows Explorer and type `%APPDATA%\Microsoft\Windows\Start Menu\Programs`. If you see a Miniconda directory, go ahead and delete it.
* Clean up environment variables. To get there, hit the Windows key and start typing "environment" and "Edit the system environment variables" should pop up. Click on that and then on the "Environment Variables" button at the bottom. Find `Path` and see if the Miinconda directory is in there. If so, edit the variable and remove it.
* Remove a few things from the registry. Open the Registry Editor by hitting the Windows key and searching for it. If you installed it for "Just me", open the `HKEY_CURRENT_USER` folder on the side panel, otherwise `HKEY_LOCAL_MACHINE`. 
  - Now go into `Software\Microsoft\Windows\CurrentVersion\Uninstall` and remove the Miniconda directory in there. This will remove it from the list of installed program in your applications list.
  - Under `Software\Python\PythonCore`. It may contain directories of the python version for your Miniconda installation. Open that directory and check `InstallPath`. If that points to the failed Miniconda installation, go ahead and remove the directory for the version.

Once all that is done, your system should be free of Miniconda traces. Go ahead and run the installer again and let me know which error you are seeing. Things that will help debugging are:

* The Miniconda version
* Whether you installing for "Just me" or "All users"
* The installation directory

**Comment by morrisseyj** [comment]
Thanks @marcoesters. Really appreciate the help. Here are the answers to your questions:

- Can't recall the exact error message on original install. 
- I don't believe i had any other miniconda instalation on the system - i do however have another python installation on the system. At some point i tried to install anaconda, but that also failed - though it successfully uninstalled. I can't find any anaconda folder in `C:\Users\[user]\AppData\Local` or in `C:\ProgramData` (which was the install location for miniconda)
- Deleted miniconda3 from `C:\ProgramData` and deleted miniconda from `C:\Users\[user]\AppData\Local` (does this mean i had installed two versions?)
- No .conda folder in `%USERPROFILE%`
- No miniconda directory in `%APPDATA%\Microsoft\Windows\Start Menu\Programs`
- miniconda not on `PATH`
- In the windows registry editor `Software\Microsoft\Windows\CurrentVersion\Uninstall` showed an Anaconda (not miniconda) directory - I deleted it
- In the windows registry editor `Software\Python\PythonCore` showed python `3.11` (along with `3.12.3` (which i can access by typing py at the command line) and `3.9`, both of which i have installed separately), i removed 3.11.

Miniconda still showing in add/remove programs (but for python `3.12.1`), but when i click to uninstall, it can't be found. Tried rebooting and this persists. Should i uninstall/remove from the registry python `3.12`? If so, how do i get `3.12.1` and leave `3.12.3` in tact?

Will hold off on reinstalling until i hear about this uninstall (and in case anything above changes your diagnosis). I assume i should get things removed before reinstalling them. 

Thanks again for the help.

**Comment by marcoesters** [comment]
> Deleted miniconda3 from C:\ProgramData and deleted miniconda from C:\Users\[user]\AppData\Local (does this mean i had installed two versions?)

Possibly. I have no observed a Miniconda directory inside `AppData\Local`.

> No miniconda directory in %APPDATA%\Microsoft\Windows\Start Menu\Programs

Since you installed into `C:\ProgramData`, you likely chose to install for "All Users". In this case, the directory could be inside `C:\ProgramData\Microsoft\Windows\Start Menu\Programs`.

> In the windows registry editor Software\Python\PythonCore showed python 3.11 (along with 3.12.3 (which i can access by typing py at the command line) and 3.9, both of which i have installed separately), i removed 3.11.

We need to be careful here. Please make sure that the directory points to a Miniconda/Anaconda directory. I suspect that `3.12.3` belongs to your other Python installation. `3.9` may also be Miniconda/Anaconda, but please check before deleting it. This key has no consequence for the Miniconda installation though. It just assigns a Python executable when you use programs like VSCode and use the terminal, so you can also leave it alone.

> Miniconda still showing in add/remove programs (but for python 3.12.1), but when i click to uninstall, it can't be found.

Yes, that is expected to fail. The entries in "add/remove programs" are just entries found in the `Uninstall` registry key and contains the command to uninstall the application. Since you removed the files, it cannot find the uninstaller.

However, that also means that we haven't found the entry yet. There are two possible locations: `HKEY_CURRENT_USER\Software\Microsoft\Windows\CurrentVersion\Uninstall` or `HKEY_LOCAL_MACHINE\Software\Microsoft\Windows\CurrentVersion\Uninstall`. Since you installed into `C:\ProgramData`, I suspect it is in `HKEY_LOCAL_MACHINE\Software\Microsoft\Windows\CurrentVersion\Uninstall`, but go ahead and check both.

> Will hold off on reinstalling until i hear about this uninstall (and in case anything above changes your diagnosis). I assume i should get things removed before reinstalling them.

Yes, that is prudent or debugging will get difficult. Since you already have another Python 3.12 installation on your system, I also suggest you use our Python 3.11 installer while we figure out what's going on: https://repo.anaconda.com/miniconda/Miniconda3-py311_24.3.0-0-Windows-x86_64.exe

**Comment by morrisseyj** [comment]
@marcoesters Thanks again for your follow up.

- No miniconda folder in `C:\ProgramData\Microsoft\Windows\Start Menu\Programs`
- Since it seems i had installed for all users navigating through the registry via `HKEY_LOCAL_MACHINE` to `Software\Microsoft\Windows\CurrentVersion\Uninstall`, i find and delete `Miniconda3 py312_24.1.2-0(Python 3.12.1 64 bit)`. Now miniconda no longer shows in the "Add/Remove" programs!
- No `Software\Python\PythonCore` via  `HKEY_LOCAL_MACHINE`
- Back on `\HKEY_CURRENT_USER\Software\Python\PythonCore` i see no link to miniconda. The only subdirectories are: `Help`, `Idle`, `IdleShortcuts`, `InstalledFeatures`, `InstallPath`, `PythonPath`. Regarding python 3.9, i know i installed that to get another library working - nothing to do with miniconda. 

I fear that what happened is that i tried installing `just for me`. When that failed, i tried again `for all users`. I appreciate now that this was not prudent.

Ok, with all of this now as it is, should I look to install miniconda again with the Python 3.11 installer, or is there some other check to run to make sure the previous mess is fully cleaned up?

Thanks again.

**Comment by marcoesters** [comment]
Yes, I think you are good to go to install the python 3.11 version now.

Since the uninstallation problem is solved, I will go ahead and close this issue. If you run into problems installing, please open a new issue so that other users can easily find it.

Please include a screenshot of the error and paste the output of the log you see during the installation (right-click on the output and copy to clipboard - this will copy the entire log). Thanks!

**Comment by morrisseyj** [comment]
Will do. Thanks so much for your help!

New issue created regarding the ongoing install issues: https://github.com/ContinuumIO/anaconda-issues/issues/13386

**Comment by jonmoore** [comment]
I think it's worth noting that miniconda has been affected by such issues since its inception, in a way that would be most unusual for a mature open-source package, let alone a package like Miniconda that is at the core of a commercial offering sold to large enterprises.  It is vanishingly rare that other packages fail in routine uninstallations in a way that requires manual cleanup of the registry, as (helpfully) recommended by the person assisting above.

Speaking as a personal user but also as the primary point of contact for a large Anaconda customer, I would greatly welcome a higher fraction of developer time being spent on making basic things like this "just work".  I see more and more developers moving away from Anaconda's tools entirely, despite the technical advantages of conda packaging, because they perceive that poetry, uv, etc. are faster, more robust, and easier to use.

**Comment by renet10** [comment]
Second @jonmoore's comment.  His comment is a year old and uninstalling via Win11 "Installed Apps" UI *still* fails, even for successful installs, and the Anaconda forum isn't helpful either.

This post appears as the top of Google search so I'm updating here.

It took a lot of searching and most suggestions were fiddly / outdated, but this is what worked for me:
1. First of all, the Win11 "Installed Apps" *does not* work (feel free to try it first), you will likely get "Failed to run pre_install" plus "Failed to run" every other script it calls
2. You need to figure out what user you installed Miniconda as, this is important.  The following instructions assume a user install, I do not know if it works for system installs, and this may work for Anaconda user installs as well since it appears also plagued with similar issues
3. Login as that user and open a Command Prompt.  DO NOT run as administrator as some posts suggest, because although the command will successfully uninstall miniconda it will leave a bunch of turds behind, including the Anaconda Prompt and the entry in Windows Installed Apps
5. Find the "Uninstall-Miniconda3.exe" file in the local user directory, usually C:\Users\<user>\miniconda3
6. Run the uninstall script

**Comment by marcoesters** [comment]
>  DO NOT run as administrator as some posts suggest, because although the command will successfully uninstall miniconda it will leave a bunch of turds behind, including the Anaconda Prompt and the entry in Windows Installed Apps

I created a few issues in the upstream repositories about this:

* https://github.com/conda/menuinst/issues/341
* https://github.com/conda/constructor/issues/997

There is indeed an assumption in the code that Miniconda installed as a user are not uninstalled as administrator. Even though I agree that there are good use cases to run an uninstallation as administrator even for single-user installations.

> "Failed to run pre_install" plus "Failed to run" every other script it calls

Logging definitely needs to be improved here: https://github.com/conda/constructor/issues/998

It's odd that the uninstaller works when called via the command line but not the UI because they are the same file. I personally can't reproduce it with multiple versions of Miniconda or Anaconda Distribution, so if you can recall any other details, that would be extremely helpful. For example, as you mentioned, if you were logged in as a different user than the user that installed Miniconda, whether you installed Miniconda into a different drive or non-default location, or if you have any anti-virus programs that could interfere.

**Comment by renet10** [comment]
I just tested it out, albeit in a VM environment instead of the main physical Win11 machine, as then I could start with a clean env that never had conda before.  Both install / uninstall worked fine.

There may be several mitigating circumstances on the physical machine:
1. For my first go at miniconda + Win, I did want to install for all users and ran the installer as admin, which failed for reasons I cannot recall.  I do not recall whether the conda install ran and gracefully terminated with an error, or whether it ran successfully and I found issues later on when I tried to use it 
2. I then uninstalled miniconda as admin and reinstalled as user
3. I'm also 75% sure that I had >1 go at installing as user, why I cannot remember.  If so I used the graphical uninstaller successfully.  For this reason I was surprised the graphical uninstaller did not work last week
4. I have been using miniconda for about 2 months, but Pytorch / CUDA do not support it well, which I found while installing CUDA after I installed miniconda3

In these 2 months I have been using miniconda3 normally but had to run many pip installs/uninstalls in working environments (never in the base env and never on the bare OS).  Fast forward to this week:

- Graphical uninstaller did not work
- I also found conda leftovers under the admin account local directory (.conda and miniconda3), likely a result of (2) above as I could not repro it in the VM

**Comment by marcoesters** [comment]
Thank you for your response! This all smells like permission issues of some sort or some kind of anti-virus program. On the VM, you probably have a lot more permissions and fewer system protections.

Can you clarify how you used the installer from the command line? Just `.\miniconda3\Uninstall-Miniconda3.exe /S`?

> In these 2 months I have been using miniconda3 normally but had to run many pip installs/uninstalls in working environments (never in the base env and never on the bare OS).

In newer versions of Miniconda, the uninstaller uninstalls each environment inside the installation directory recursively (via `_conda.exe remove --all`). But it seems unlikely that you wouldn't be able to run this command and run into permissions problems unless `pip install` installed files you don't have write access to or that `conda` isn't able to remove for other reasons.

The uninstaller also executes a few `pythonw.exe` calls. Most of the dialogs that pop up during the uninstallation are because of these commands failing. I've seen some security programs block them recently. When run in the command line via `Uninstall-Miniconda3.exe /S`, these commands fail silently and continue.

> I also found conda leftovers under the admin account local directory (.conda and miniconda3), likely a result of (2) above as I could not repro it in the VM

If the uninstallation fails or is aborted, those files are expected to be left over. The `.conda` directory may remain after a successful uninstallation, too, if you didn't opt into deleting the user data. That directory is created by `conda` and contains, among other files, the `environment.txt` file, which contains a list of all environments of all `conda` installations. That's why it's not removed by default.

**Comment by renet10** [comment]
I also suspect some form of permission issue.  I was wondering if perhaps CUDA installation may have botched something, especially as I have to enable Developer Mode to use it.  

To uninstall, I typed ".\miniconda3\Uninstall-Miniconda3.exe" with no flags, and I likely used cmd.exe and not PowerShell since I found PowerShell was blocking scripts unless I explicitly "Set-ExecutionPolicy".  In fact I was planning to see if Developer Mode / ExecutionPolicy was the issue but haven't had the time.

I also don't believe the Python commands failed silently, since the end results between UI-basd uninstall and command line were completely different.  The UI-based failure left everthing intact (AFAIK it did nothing) whereas the command line cleaned everything up nicely.

Also in case it is significant, the Python version I was using Conda with was 3.13, but when I switched to venv I also installed 3.12 just before I tried to uninstall Miniconda3.

I'm happy to see if I can repro it again and get logs, if you can point me at how to get the logs.

Thanks for the all the support.  To be clear conda was great while I was using it but PyTorch/CUDA forced the switch.

**Comment by marcoesters** [comment]
>  I was wondering if perhaps CUDA installation may have botched something, especially as I have to enable Developer Mode to use it.

The only way I can see that happening is if a package created a file without you having write permissions, which would strike me as odd.

> To uninstall, I typed ".\miniconda3\Uninstall-Miniconda3.exe" with no flags

That would mean that the GUI installer was used in either case since calling the installer in the command line without `/S` just opens the graphical installer. Since nothing was uninstalled when called from the Windows UI, that would almost suggest that the user that calls the installer from the Windows UI is different than the user that opened the uninstaller from the command line. That would be very strange though.

> if you can point me at how to get the logs.

The text in the white box that you see during the uninstallation can be copied to the clipboard. That's the only logs you can get for now.

**Comment by renet10** [comment]
I just tried enabling Developer Mode in my VM env and ofc miniconda3 uninstalled successfully, as I'm sure you expected.  I'm as mystified as you are.

The only thing I personally did was install CUDA which installs way too much other stuff, but it's also the gaming machine so other things may have happened without me knowing.  Short of that, it was using conda as normal.

Regarding the failure messages, the symptoms I saw mirrored [this post by someone else on StackExchange](https://stackoverflow.com/questions/78106430/uninstalling-anaconda-failed-to-run-pre-uninstall).
